tango/integrations/gs/workspace.py

```diff
@@ -0,0 +1,225 @@
+        return cls(workspace)
+
+    @property
+    def url(self) -> str:
+        return self.client.url()
+
+    def _remote_lock(self, step: Step) -> GCSStepLock:
+        return GCSStepLock(self.client, step)
+
+    def _step_location(self, step: Step) -> str:
+        return self.client.url(self.Constants.step_dataset_name(step))
+
+    def _save_run(
+        self, steps: Dict[str, StepInfo], run_data: Dict[str, str], name: Optional[str] = None
+    ) -> Run:
+        if name is None:
+            while True:
+                name = petname.generate() + str(random.randint(0, 100))
+                if not self._ds.get(self._ds.key("run", name)):
+                    break
+        else:
+            if self._ds.get(self._ds.key("run", name)):
+                raise ValueError(f"Run name '{name}' is already in use")
+
+        run_entity = self._ds.entity(key=self._ds.key("run", name), exclude_from_indexes=("steps",))
+        run_entity["start_date"] = datetime.datetime.now()
+        run_entity["steps"] = json.dumps(run_data).encode()
+        self._ds.put(run_entity)
+
+        return Run(name=cast(str, name), steps=steps, start_date=run_entity["start_date"])
+
+    def _get_run_from_entity(self, run_entity: datastore.Entity) -> Optional[Run]:
+        try:
+            steps_info_bytes = run_entity["steps"]
+            steps_info = json.loads(steps_info_bytes)
+        except KeyError:
+            return None
+
+        import concurrent.futures
+
+        steps: Dict[str, StepInfo] = {}
+        with concurrent.futures.ThreadPoolExecutor(
+            max_workers=self.NUM_CONCURRENT_WORKERS,
+            thread_name_prefix="GSWorkspace._get_run_from_dataset()-",
+        ) as executor:
+            step_info_futures = []
+            for unique_id in steps_info.values():
+                step_info_futures.append(executor.submit(self.step_info, unique_id))
+            for future in concurrent.futures.as_completed(step_info_futures):
+                step_info = future.result()
+                assert step_info.step_name is not None
+                steps[step_info.step_name] = step_info
+
+        return Run(name=run_entity.key.name, start_date=run_entity["start_date"], steps=steps)
+
+    def registered_runs(self) -> Dict[str, Run]:
+        import concurrent.futures
+
+        runs: Dict[str, Run] = {}
+
+        with concurrent.futures.ThreadPoolExecutor(
+            max_workers=self.NUM_CONCURRENT_WORKERS,
+            thread_name_prefix="GSWorkspace.registered_runs()-",
+        ) as executor:
+            run_futures = []
+            for run_entity in self._ds.query(kind="run").fetch():
+                run_futures.append(executor.submit(self._get_run_from_entity, run_entity))
+            for future in concurrent.futures.as_completed(run_futures):
+                run = future.result()
+                if run is not None:
+                    runs[run.name] = run
```

Comment:
So run info is still in GCS, not in Datastore. If it's fast enough, even with many runs, that's fine. But it might be a good candidate to put into Datastore, too.

Reply:
Run info _is_ in datastore (`self._ds` is the datastore). The only run-related information stored in GCS is the log file.